test/libsolidity/semanticTests/inlineAssembly/transient_storage_selfdestruct.sol

```diff
@@ -0,0 +1,60 @@
+contract C {
+    uint8 x;
+
+  constructor() payable {
+    x = 8;
+    assembly {
+        tstore(0, 13)
+    }
+  }
+
+  function getTransientStorage() external view returns (uint y) {
+    assembly {
+        y := tload(0)
+    }
+  }
+
+  function getX() external view returns (uint8) {
+    return x;
+  }
+
+  function terminate() external {
+    selfdestruct(payable(msg.sender));
+    assert(false);
+  }
+}
+
+contract D {
+  C public c;
+
+  constructor() payable {
+      c = new C();
+  }
+
+  function f() external {
+      c.terminate();
+      assert(c.getX() == 8);
+      assert(c.getTransientStorage() == 0);
+      assert(this.exists() == true);
+  }
+
+  function g() external {
+    C c2 = new C();
+    c2.terminate();
+    assert(c2.getX() == 8);
```

Comment:
@cameel , I tried to implement the test with `selfdestruct`, but shouldn't `storage` be cleaned as well? Not sure if I did something wrong in the test case, but looks like neither `storage` nor `transient storage` data are being deleted.

Reply:
Haha, yeah, I just noticed that too: https://github.com/ethereum/solidity/pull/14737#discussion_r1462320683

I assumed storage would be cleared immediately, but looks like it only gets cleared at the end of the transaction, just like code. I checked execution specs and it confirms it - [`selfdestruct` simply adds the account to the `accounts_to_delete` list](https://github.com/ethereum/execution-specs/blob/forks/cancun/src/ethereum/cancun/vm/instructions/system.py#L525-L526), and it's only at the end of transaction that [this list is iterated and accounts destroyed](https://github.com/ethereum/execution-specs/blob/forks/cancun/src/ethereum/cancun/fork.py#L644-L645).

Also, apparently [transient storage is not touched at all when the account is destroyed, only storage is](https://github.com/ethereum/execution-specs/blob/forks/cancun/src/ethereum/cancun/state.py#L191-L192). Could be that this spec is incomplete there (it's still in a branch) but more likely that it's not necessary because it's implicitly cleared at the end of transactions.

So yeah, I think we can mark `selfdestruct` as not having any effect on transient storage.

Not sure about storage - it seems to me that `selfdestruct` should be marked as writing to it - but even if it's wrong, it must have been wrong for a long time now and we should deal with that separately, not in this PR. It may actually be fine since `selfdestruct` is so special - it writes everything and does not read anything.

Reply:
Done.